File: gfx2Imp/js/FwtGraphics2.js
//
// Copyright (c) 2011, chunquedong
// Licensed under the Academic Free License version 3.0
//
// History:
//   2011-08-20  Jed Young  Creation
//

fan.fwt.Graphics.prototype.drawImage2 = function(image, x, y)
{
  this.cx.drawImage(image.getImage(), 0, 0);
  return this;
}

fan.fwt.Graphics.prototype.copyImage2 = function(Image2 image, Rect src, Rect dest)
{
  var jsImg = fan.fwt.FwtEnvPeer.loadImage(fanImg);
  if (jsImg.width > 0 && jsImg.height > 0)
    this.cx.drawImage(jsImg, src.m_x, src.m_y, src.m_w, src.m_h, dst.m_x, dst.m_y, dst.m_w, dst.m_h)
  return this;
}

fan.fwt.Graphics.prototype.drawPath = function(path)
{
  fan.fwt.Graphics.doJsPath(this.cx, path);
  this.cx.stroke();
  return this;
}

fan.fwt.Graphics.prototype.fillPath = function(path)
{
  fan.fwt.Graphics.doJsPath(this.cx, path);
  this.cx.fill();
  return this;
}

fan.fwt.Graphics.prototype.drawPolyline2 = function(p)
{
  this.cx.beginPath();
  var size = p.size();
  var x;
  var y;
  for (var i=0; i < size; i+=2)
  {
    x = p.get(i);
    y = p.get(i+1);
    if (i == 0) this.cx.moveTo(x, y);
    else this.cx.lineTo(x, y);
  }
  this.cx.stroke();
  return this;
}

fan.fwt.Graphics.prototype.fillPolygon2 = function(p)
{
  this.cx.beginPath();
  var size = p.size();
  var x;
  var y;
  for (var i=0; i < size; i+=2)
  {
    x = p.get(i);
    y = p.get(i+1);
    if (i == 0) this.cx.moveTo(x, y);
    else this.cx.lineTo(x, y);
  }
  this.cx.closePath();
  this.cx.fill();
  return this;
}

fan.fwt.Graphics.prototype.setTransform = function(trans)
{
  fan.fwt.Graphics.doJsTransform(this.cs, trans);
  return this;
}

fan.fwt.Graphics.prototype.setClipping = function(path)
{
  fan.fwt.Graphics.doJsPath(this.cx, path);
  this.cx.clip();
  return this;
}


//////////////////////////////////////////////////////////////////////////
// Util
//////////////////////////////////////////////////////////////////////////

fan.fwt.Graphics.doJsPath = function(cx, path)
{
  var size = path.steps().size();
  cx.beginPath();

  for (var i =0; i < size; ++i)
  {
    var step = path.steps().get(i);

    if (step instanceof fan.gfx2.PathMoveTo)
    {
      cx.moveTo(s.x, s.y);
    }
    else if (step instanceof fan.gfx2.PathLineTo)
    {
      cx.lineTo(s.x, s.y);
    }
    else if (step instanceof fan.gfx2.PathQuadTo)
    {
      cx.quadraticCurveTo(s.cx, s.cy, s.x, s.y);
    }
    else if (step instanceof fan.gfx2.PathCubicTo)
    {
      cx.bezierCurveTo(s.cx1, s.cy1, s.cx2, s.cy2, s.x, s.y);
    }
    else if (step instanceof fan.gfx2.PathClose)
    {
      cs.closePath();
    }
    else
    {
      throw fan.sys.Err.make("unreachable");
    }
  }
}

fan.fwt.Graphics.doJsTransform = function(cx, trans)
{
  cx.setTransform(
       trans.get(0,0),
       trans.get(1,0),
       trans.get(0,1),
       trans.get(1,1),
       trans.get(2,0),
       trans.get(2,1)
     );
}

File: gfx2Imp/java/FwtGraphics2.java
//
// Copyright (c) 2011, chunquedong
// Licensed under the Academic Free License version 3.0
//
// History:
//   2011-08-14  Jed Young  Creation
//

package fan.gfx2Imp;

import java.util.Stack;

import fan.sys.FanObj;
import fan.sys.ArgErr;

import fan.array.*;
import fan.fan3dMath.*;
import fan.gfx.*;
import fan.gfx2.*;
import fan.fwt.FwtGraphics;


import org.eclipse.swt.SWT;
import org.eclipse.swt.graphics.GC;
import org.eclipse.swt.graphics.Pattern;
import org.eclipse.swt.graphics.Rectangle;
import org.eclipse.swt.graphics.Transform;
import org.eclipse.swt.widgets.Display;
import org.eclipse.swt.events.PaintEvent;

public class FwtGraphics2 extends FwtGraphics implements Graphics2
{
  GC gc;

  public FwtGraphics2(PaintEvent e)
  {
    this(e.gc, e.x, e.y, e.width, e.height);
  }

  public FwtGraphics2(GC gc, int x, int y, int w, int h)
  {
    super(gc, x, y, w, h);
    this.gc = gc;
  }

  public FwtGraphics2 drawImage2(Image2 image, long x, long y)
  {
    PixmapImp p = (PixmapImp)image;
    gc.drawImage(p.getImage(), (int)x, (int)y);
    return this;
  }
  public FwtGraphics2 copyImage2(Image2 image, Rect s, Rect d)
  {
    PixmapImp p = (PixmapImp)image;
    gc.drawImage(p.getImage(),
      (int)s.x, (int)s.y, (int)s.w, (int)s.h,
      (int)d.x, (int)d.y, (int)d.w, (int)d.h);
    return this;
  }

  public FwtGraphics2 drawPath(fan.gfx2.Path path)
  {
    gc.drawPath(toSwtPath(path));
    return this;
  }
  public FwtGraphics2 fillPath(fan.gfx2.Path path)
  {
    gc.fillPath(toSwtPath(path));
    return this;
  }

  public FwtGraphics2 drawPolyline2(Array p)
  {
    gc.drawPolyline((int[])p.peer.getValue());
    return this;
  }
  public FwtGraphics2 fillPolygon2(Array p)
  {
    gc.fillPolygon((int[])p.peer.getValue());
    return this;
  }

  public FwtGraphics2 setTransform(fan.fan3dMath.Transform2D t)
  {
    gc.setTransform(toSwtTransform(t));
    return this;
  }

  public FwtGraphics2 clipPath(Path path)
  {
    gc.setClipping(toSwtPath(path));
    return this;
  }

  public Pixmap image()
  {
    return new PixmapImp(gc.getGCData().image);
  }

  /**
   * auto free resource
   */
  @Override
  protected void finalize()
  {
    if (!gc.isDisposed()) gc.dispose();
  }


//////////////////////////////////////////////////////////////////////////
// Util
//////////////////////////////////////////////////////////////////////////

  static public org.eclipse.swt.graphics.Path toSwtPath(fan.gfx2.Path path)
  {
    int size = (int)path.steps().size();
    org.eclipse.swt.graphics.Path swtPath = new org.eclipse.swt.graphics.Path(FwtEnv2Peer.getDisplay());
    for (int i =0; i < size; ++i)
    {
      PathStep step = (PathStep)path.steps().get(i);

      if (step instanceof PathMoveTo)
      {
        PathMoveTo s = (PathMoveTo)step;
        swtPath.moveTo((float)s.x, (float)s.y);
      }
      else if (step instanceof PathLineTo)
      {
        PathLineTo s = (PathLineTo)step;
        swtPath.lineTo((float)s.x, (float)s.y);
      }
      else if (step instanceof PathQuadTo)
      {
        PathQuadTo s = (PathQuadTo)step;
        swtPath.quadTo((float)s.cx, (float)s.cy, (float)s.x, (float)s.y);
      }
      else if (step instanceof PathCubicTo)
      {
        PathCubicTo s = (PathCubicTo)step;
        swtPath.cubicTo((float)s.cx1, (float)s.cy1, (float)s.cx2, (float)s.cy2, (float)s.x, (float)s.y);
      }
      else if (step instanceof PathClose)
      {
        swtPath.close();
      }
      else
      {
        throw fan.sys.Err.make("unreachable");
      }
    }
    return swtPath;
  }

  static public Transform toSwtTransform(Transform2D trans)
  {
    return new Transform(FwtEnv2Peer.getDisplay(),
       (float)trans.get(0,0),
       (float)trans.get(1,0),
       (float)trans.get(0,1),
       (float)trans.get(1,1),
       (float)trans.get(2,0),
       (float)trans.get(2,1)
       );
  }
}